bluetooth: Remove useless LE Set Data Length after Connected

rugeGerritsen opened the pull request:
Calling LE Write Suggested Default Data Length command is sufficient
to ensure that the LL will auto-initiate the DLE procedure. Performing
a second procedure is unecessary.

As some controllers don't follow the spec recommendation, a controller quirk is added

In Core v5.2, Vol 4, Part E, Section 7.8.35:
_The HCI_LE_Write_Suggested_Default_Data_Length command allows the
Host to specify its suggested values for the Controller's maximum
transmission number of payload octets and maximum packet
transmission time for packets containing LL Data PDUs to be used
for new connections. The Controller may use smaller or larger values
for connInitialMaxTxOctets and connInitialMaxTxTime based on local
information._

In short, this command sets connInitialMaxTxOctets
and connInitialMaxTxTime.

In Core v5.2, Vol 6, Part B, Section 4.5.10:
_For a new connection: connMaxTxOctets shall be set to
connInitialMaxTxOctets and connMaxRxOctets shall be chosen by the
Controller. If either value is not 27 then the Controller should
initiate the Data Length Update Procedure at the earliest practical
opportunity._

Therefore, calling LE Write Suggested Default Data Length command
is sufficient to ensure the DLE being performed.

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

rugeGerritsen commented:
@cvinayak . You may be interested in this as well. The Zephyr controller does not auto-initiate the DLE procedure as required by spec.

cvinayak commented:
@rugeGerritsen its a "should" not a "shall".

joerchan commented:
@rugeGerritsen We need to remove the conn state flag `BT_CONN_AUTO_DATA_LEN_COMPLETE` it will prevent the user controlled data len procedure to be called now.

rugeGerritsen commented:
> @rugeGerritsen its a "should" not a "shall".

Thanks for pointing out. In either case, it would be nice if the host assumes the behavior recommended by the specification.

rugeGerritsen commented:
> @rugeGerritsen We need to remove the conn state flag `BT_CONN_AUTO_DATA_LEN_COMPLETE` it will prevent the user controlled data len procedure to be called now.

Done

joerchan commented:
@carlescufi @rugeGerritsen @cvinayak Shall (should?) we come to an agreement here?
The question is if we would want to move this handling to zephyr controller or not.
The host should assume that the HCI controller uses the recommended practice, that the LE Write Default DLE/PHY would auto-initiate the procedure in the controller. 
Maybe we could put sending the LE Set Data Len as an HCI driver quirk?

carlescufi commented:
> @carlescufi @rugeGerritsen @cvinayak Shall (should?) we come to an agreement here?
> The question is if we would want to move this handling to zephyr controller or not.
> The host should assume that the HCI controller uses the recommended practice, that the LE Write Default DLE/PHY would auto-initiate the procedure in the controller.
> Maybe we could put sending the LE Set Data Len as an HCI driver quirk?

I am in favor of delegating this to the controller, although it will require a bit of an effort. In the meantime I think adding a quirk (or a Kconfig option) seems like the path forward in order to be compatible with both kinds of controllers.
I have also added @kruithofa to the reviewer list so that he can discuss with @cvinayak how best to implement this in the controller.

rugeGerritsen commented:
@cvinayak , @kruithofa , do you have input to this issue?

joerchan commented:
@rugeGerritsen 

> I agree that it would be best to implement this in the controller. I believe it is quite simple to implement.
> 
> We still may need the configurability in the host. What about the following proposal: Split the CONFIG_BT_AUTO_DATA_LEN_UPDATE to CONFIG_BT_SET_DEFAULT_DATA_LEN and CONFIG_BT_DATE_LEN_UPDATE_ON_CONN_COMPLETE. We can then add overlays or something similar for the different controllers. A similar approach can be done for the PHY procedure.

I don't think we should have this as Kconfig, as it is more of a strangeness than a feature. I think either BT_QUIRK_ (hci_driver.h) or maybe just wrap that code in ifdef LL_SW_SPLIT until we have it fixed in the controller.

rugeGerritsen commented:
> I don't think we should have this as Kconfig, as it is more of a strangeness than a feature. I think either BT_QUIRK_ (hci_driver.h) or maybe just wrap that code in ifdef LL_SW_SPLIT until we have it fixed in the controller.

Wrapping it in ifdef LL_SW_SPLIT won't work in multicore systems. We therefore need to implement it either as a quick or in the controller.

joerchan commented:
I would add a quirk, and then remove the whole the quirk again when it is implemented by the controller.

rugeGerritsen committed:
bluetooth: hci_driver: Add BT_QUIRK_NO_AUTO_DLE

Some controllers do not follow the recommendation of auto-initiating
a data length update if the host writes default data length parameters.

If the controller follows this recommendation, calling
LE Write Suggested Default Data Length command is sufficient
to ensure that the LL will auto-initiate the DLE procedure. Performing
a second procedure is unecessary.

In Core v5.2, Vol 4, Part E, Section 7.8.35:
The HCI_LE_Write_Suggested_Default_Data_Length command allows the
Host to specify its suggested values for the Controller's maximum
transmission number of payload octets and maximum packet
transmission time for packets containing LL Data PDUs to be used
for new connections. The Controller may use smaller or larger values
for connInitialMaxTxOctets and connInitialMaxTxTime based on local
information.

In short, this command sets connInitialMaxTxOctets
and connInitialMaxTxTime.

In Core v5.2, Vol 6, Part B, Section 4.5.10:

For a new connection: connMaxTxOctets shall be set to
connInitialMaxTxOctets and connMaxRxOctets shall be chosen by the
Controller. If either value is not 27 then the Controller should
initiate the Data Length Update Procedure at the earliest practical
opportunity.

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

rugeGerritsen committed:
bluetooth: controller: Add BT_QUIRK_NO_AUTO_DLE for BT_LL_SW_SPLIT

It does not auto-initiate the DLE procedure if the default data length
parameters are not equal to the initial parameters.

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -1941,15 +1941,21 @@ static void conn_auto_initiate(struct bt_conn *conn)
 	if (IS_ENABLED(CONFIG_BT_AUTO_DATA_LEN_UPDATE) &&
-	    BT_FEAT_LE_DLE(bt_dev.le.features)) {
-		uint16_t tx_octets, tx_time;
+		BT_FEAT_LE_DLE(bt_dev.le.features)) {
```
align with rest of if-statement.

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -2320,7 +2326,8 @@ static void le_data_len_change(struct net_buf *buf)
 #if defined(CONFIG_BT_USER_DATA_LEN_UPDATE)
-	if (IS_ENABLED(CONFIG_BT_AUTO_DATA_LEN_UPDATE)) {
+	if (IS_ENABLED(CONFIG_BT_AUTO_DATA_LEN_UPDATE) &&
+		(bt_dev.drv->quirks & BT_QUIRK_NO_AUTO_DLE)) {
```
align with rest of if-statement

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -1941,15 +1941,21 @@ static void conn_auto_initiate(struct bt_conn *conn)
+		} else {
+			/* No need to auto-initiate DLE procedure.
+			 * It is done by the controller. */
+			atomic_set_bit(conn->flags, BT_CONN_AUTO_DATA_LEN_COMPLETE);
```
Wouldn't it be enough to only set this in the le_data_len_change?
Now the application can start the user initiated data length update procedure while the controller initiated data length update might be running.

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5688,28 +5695,34 @@ static int le_init(void)
-		struct bt_hci_cp_le_write_default_data_len *cp;
-		uint16_t tx_octets, tx_time;
+		if ((IS_ENABLED(CONFIG_BT_AUTO_DATA_LEN_UPDATE) &&
+			!(bt_dev.drv->quirks & BT_QUIRK_NO_AUTO_DLE))) {
```
align with the rest of the if-statement.

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5688,28 +5695,34 @@ static int le_init(void)
+				return err;
+			}
+		} else {
+			/* No need to set default data length.
```
Comment is a misleading if CONFIG_BT_AUTO_DATA_LEN_UPDATE is disabled

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -170,7 +170,7 @@ NET_BUF_POOL_FIXED_DEFINE(hci_rx_pool, CONFIG_BT_RX_BUF_COUNT,
 
 #if defined(CONFIG_BT_CONN)
-#define NUM_COMLETE_EVENT_SIZE BT_BUF_SIZE(                            \
+#define NUM_COMLETE_EVENT_SIZE BT_BUF_SIZE(\
```
accidental change here?

joerchan reviewed (approved):
Looks good, thank you.

cvinayak commented:
> 
> 
> @cvinayak , @kruithofa , do you have input to this issue?

The auto-initiation in the current controller LLCP, not going to invest. In the new LLCP, yes, can be added as select-able feature.

joerchan commented:
>  In the new LLCP, yes, can be added as select-able feature.

I think it is better to just make that only behavior, otherwise what is the point of the set default command?

cvinayak commented:
> 
> 
> > In the new LLCP, yes, can be added as select-able feature.
> 
> I think it is better to just make that only behavior, otherwise what is the point of the set default command?

The values from the set default length command will get used in the length response when the auto-initiation is disabled. As the host does not auto-initiate ATT MTU request, or would synchronize with link layer as to when length request is sent, it should be at the applications discretion to optimally use these procedure. 

Both data length and phy update procedures where auto-initiated by default in host, and it has been now made as Kconfig configurable.

joerchan commented:
@cvinayak Ok. Do you have any comments to this PR and the way the host now uses these commands?

cvinayak commented on include/drivers/bluetooth/hci_driver.h:
```diff
@@ -30,6 +30,11 @@ extern "C" {
+	 * the DLE procedure after connection establishment. */
+	BT_QUIRK_NO_AUTO_DLE = BIT(1),
 };
 
```
```suggestion
#define IS_BT_QUIRK_NO_AUTO_DLE(bt_dev) ((bt_dev)->drv.quirks & BT_QUIRK_NO_AUTO_DLE)
```

cvinayak commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -1942,14 +1942,21 @@ static void conn_auto_initiate(struct bt_conn *conn)
 	if (IS_ENABLED(CONFIG_BT_AUTO_DATA_LEN_UPDATE) &&
 	    BT_FEAT_LE_DLE(bt_dev.le.features)) {
-		uint16_t tx_octets, tx_time;
+		if (bt_dev.drv->quirks & BT_QUIRK_NO_AUTO_DLE) {
```
```suggestion
		if (IS_BT_QUIRK_NO_AUTO_DLE(&bt_dev)) {
```

cvinayak reviewed (changes requested):
Minor nitpick on use of a check macro; Otherwise, changes look good to me.

rugeGerritsen commented:
Rebased to get rid of the conflict in hci_driver.h

joerchan commented:
@jhedberg Can you have a look?

carlescufi commented:
Ops, I thought this was reasonable so I merged, but I failed to see that @joerchan had asked @jhedberg to take a look. @jhedberg sorry about that, let  me know if you have any objections.
